src/drivers/sifive_spi0.c

```diff
@@ -4,92 +4,100 @@
 #include <metal/machine/platform.h>
 
 #ifdef METAL_SIFIVE_SPI0
-
 #include <metal/drivers/sifive_spi0.h>
 #include <metal/io.h>
 #include <metal/machine.h>
+#include <time.h>
 
 /* Register fields */
 #define METAL_SPI_SCKDIV_MASK         0xFFF
 
 #define METAL_SPI_SCKMODE_PHA_SHIFT   0
 #define METAL_SPI_SCKMODE_POL_SHIFT   1
 
-#define METAL_SPI_CSMODE_MASK         0x3
+#define METAL_SPI_CSMODE_MASK         3
 #define METAL_SPI_CSMODE_AUTO         0
 #define METAL_SPI_CSMODE_HOLD         2
 #define METAL_SPI_CSMODE_OFF          3
 
-#define METAL_SPI_PROTO_MASK          0x3
+#define METAL_SPI_PROTO_MASK          3
 #define METAL_SPI_PROTO_SINGLE        0
 #define METAL_SPI_PROTO_DUAL          1
 #define METAL_SPI_PROTO_QUAD          2
 
-#define METAL_SPI_ENDIAN_LSB          (1 << 2)
+#define METAL_SPI_ENDIAN_LSB          4
 
-#define METAL_SPI_DISABLE_RX          (1 << 3)
+#define METAL_SPI_DISABLE_RX          8
 
 #define METAL_SPI_FRAME_LEN_SHIFT     16
-#define METAL_SPI_FRAME_LEN_MASK      (0x1F << METAL_SPI_FRAME_LEN_SHIFT)
+#define METAL_SPI_FRAME_LEN_MASK      (0xF << METAL_SPI_FRAME_LEN_SHIFT)
 
 #define METAL_SPI_TXDATA_FULL         (1 << 31)
 #define METAL_SPI_RXDATA_EMPTY        (1 << 31)
-#define METAL_SPI_TXMARK_MASK         0x3
-#define METAL_SPI_TXWM                (1 << 0)
+#define METAL_SPI_TXMARK_MASK         7
+#define METAL_SPI_TXWM                1
+#define METAL_SPI_TXRXDATA_MASK       (0xFF)
+
+#define METAL_SPI_INTERVAL_SHIFT      16
+
+#define METAL_SPI_CONTROL_IO          0
+#define METAL_SPI_CONTROL_MAPPED      1
 
 #define METAL_SPI_REG(offset)   (((unsigned long)control_base + offset))
 #define METAL_SPI_REGB(offset)  (__METAL_ACCESS_ONCE((__metal_io_u8  *)METAL_SPI_REG(offset)))
 #define METAL_SPI_REGW(offset)  (__METAL_ACCESS_ONCE((__metal_io_u32 *)METAL_SPI_REG(offset)))
 
+#define METAL_SPI_RXDATA_TIMEOUT      1
+
 static int configure_spi(struct __metal_driver_sifive_spi0 *spi, struct metal_spi_config *config)
 {
     long control_base = __metal_driver_sifive_spi0_control_base((struct metal_spi *)spi);
     /* Set protocol */
-    METAL_SPI_REGW(METAL_SIFIVE_SPI0_FMT) &= ~(METAL_SPI_PROTO_MASK);
+    METAL_SPI_REGB(METAL_SIFIVE_SPI0_FMT) &= ~(METAL_SPI_PROTO_MASK);
```

Comment:
Wait. `METAL_SPI_REGB()` only changes a single byte. I was suggesting it exclusively in the context of `txdata` which only needs to update the least-significant byte, but you really should use `REGW` when you're trying to update an entire word.